00XX-efficient-msms-for-non-native-pickles-verification.md

```diff
@@ -0,0 +1,483 @@
+        (\overbrace{G_i 2^{j \cdot k}}^{\text{Coordinates in $\mathbb{F}_{base}(\mathrm{Vesta})$, computed externally}}
+        )\Big)}_{\text{Encoded in $\mathbb{F}_{scalar}(\textrm{BN254})$}} \\
+&= \sum_{j=1}^{l} \Big(\sum_{i=1}^n \underbrace{c_{i,j}}_{\text{Coefficients for limb $j$}} \cdot \overbrace{(2^{j \cdot k} \cdot G_i )}^{\text{"Bases" for limb $j$}}\Big) \\
+\end{align*}
+```
+
+<!-- Note that we have now N * l base elements, not only N -->
+
+The coefficients $c_{i, j}$ will be encoded on $2^k$ bits, with $k$ small compared to the field size (around 15). The scaled base elements $2^{j \cdot k} \cdot G_{i}$ can be pre-computed outside of the circuit.
+
+In the rest of the section we describe the MSM algorithm that efficiently computes the inner sum. The main strength of the MSM algorithm is that due to coefficients being small we can use RAM lookups on `buckets` which speeds up things quite a bit. The MSM algorithm is implementing a standard 'bucketing' trick with $2^k$ buckets to avoid doing any doublings or scalings at all, requiring only $l \cdot n + 2^{k+1}$ curve point additions.
+
+Note how in the formula above we can iterate over all the possible coefficients $c_{i,j}$ without knowing the indices in advance -- a coefficient $c$ and limb number $j$ defines positions $i$ in which the index is used. For a given $c$, let $\hat G_{c,j}$ be the value equal to $\sum_i 2^{j \cdot k} \cdot G_i$ for all $c_{i,j}$ (coefficient number $i$ for limb number $j$) equal to $c$. Then the previous decomposition equation can be expressed as:
+
+```math
+\begin{align*}
+\sum_{i=1}^n c_i G_i &= \sum_{j=1}^{l} \Big(\sum_{c \in [0,2^k-1]} c \cdot \hat G_{c,j}\Big) \\
+&= \sum_{c \in [0,2^k-1]} c \cdot (\sum_{j=1}^{l} \hat G_{c,j})
+\end{align*}
+```
+
+This suggests that we can first accumulate the values of $`\hat G_{c,j}`$ into the correct "buckets", where each bucket corresponds to a coefficient $c_i \in [0,2^k-1]$, and then sum all of them scaling by $c_{i}$.
+
+A lookup table will be used to fetch the corresponding $G_i \cdot 2^{j \cdot k}$. Therefore, the only operations that we need to encoded is the addition of $`\mathbb{F}_{base}(\mathrm{Vesta})`$ elements in $`\mathbb{F}_{scalar}(\mathrm{BN254})`$. Note that the elements $G_i \cdot 2^{j \cdot k}$ will have coordinates in $`\mathbb{F}_{base}(\mathrm{Vesta})`$. Therefore, the table will require more than one limbs for each coordinates.
+
+Assuming the additive notation for the group, the formal code description of the algorithm is as follows:
+
+```rust
+fn fill_buckets(coeffs: Vec<Field>, bases: Vec<Field>, k: uint, H: Group) {
+    for (coefficient, commitment) in coeffs.iter().zip(bases.iter()) {
+        buckets[coefficient] += commitment;
+    }
+}
+
+/// Computes the total MSM: \sum_{i=1}^N coeffs_i \cdot bases_i
+fn compute_msm(coeffs: Vec<Field>, bases: Vec<Group>, k: uint, H: Group) {
+    // Initialize the buckets with the blinding factor H
```

Comment:
It might have been my mis-wording, but: `H` is the blinding element, to which the blinding factor is applied.